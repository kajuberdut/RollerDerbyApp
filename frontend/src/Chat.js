import React, { useState, useEffect } from "react";
import io from "socket.io-client";
import {
    Card,
    CardBody,
    CardTitle,
    Form,
    FormGroup,
    Label, 
    Input,
    Button,
  } from "reactstrap";

  
// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!! yours is below !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

const Chat = () => {
    const [messages, setMessages] = useState([]);
    const [currentMessage, setCurrentMessage] = useState('');
    // const [userId, setUserId] = useState(Date.now());
    const [userId, setUserId] = useState(2);
    let INITIAL_STATE = { message: ""};
  /** Sets formData in initial state */
    const [formData, setFormData] = useState(INITIAL_STATE);
    const [isConnected, setIsConnected] = useState(false);
    const [socket, setSocket] = useState();
    


    
    // console.log("before web socket in chat.js")
    // const socket = new WebSocket(`ws://localhost:8000/ws/${userId}`);
    // console.log("socket: ", socket)

    useEffect(() => {
      console.log("before web socket in chat.js")
      const socket = new WebSocket(`ws://localhost:8000/ws/${userId}`);
      console.log("socket: ", socket)
      // Add event listeners for socket events (e.g., onmessage, onopen, onclose)
      setIsConnected(true)

      setSocket(socket)
      // Cleanup function to close the socket when the component unmounts
      return () => socket.close();
  }, []);
  

    const handleSubmit = async evt => {
        evt.preventDefault();   
        console.log("formData in handle Submit chat.js:", formData)
        // ! this is breaking the app but doesnt start sending requests and get infinite 404 requests until it is 
        // const socket = io('http://localhost:8000', {path: `/ws/${userId}`});
        
        socket.send(formData.message)
        console.log("No error?")
        }

    const handleChange = evt => {
        console.log('handleChange is running')
    
        const { name, value }= evt.target;
    
        setFormData(fData => ({
          ...fData,
          [name]: value,
          
        }));
        console.log("formData in chat.js:", formData)
      };
  
    return (
        <div className="" style={{marginTop: "150px"}}>
            <h2>status: {isConnected ? 'connected' : 'disconnected'}</h2>
            <div style={{height: '500px', overflowY: 'scroll', border: 'solid black 2px', padding: '10px', marginTop: '15px', display: 'flex', flexDirection: 'column'}}>
                <Form onSubmit={handleSubmit}>
                <Label htmlFor="message">Message: </Label>
                       
                       <Input
                           type="textarea"
                           id="message"
                           name="message"
                           value={formData.message}
                           onChange={handleChange}
                           placeholder="Type a message.."
                       />
                       <Button>Send</Button>
                </Form>
      
              </div>
            </div>
              
    )
};
  
export default Chat;



// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!! yours is above !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        
        //  <head>
        //          <title>Websocket Demo</title>
        //             <!-- Bootstrap CSS -->
        //      <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-EVSTQN3/azprG1Anm3QDgpJLIm9Nao0Yz1ztcQTwFspd3yD65VohhpuuCOmLASjC" crossorigin="anonymous">
        
        //      </head>
        //      <body>
        //      <div class="container mt-3">
        //          <h1>FastAPI WebSocket Chat</h1>
        //         <h2>Your ID: <span id="ws-id"></span></h2>
        //          <form action="" onsubmit="sendMessage(event)">
        //              <input type="text" class="form-control" id="messageText" autocomplete="off"/>
        //              <button class="btn btn-outline-primary mt-2">Send</button>
        //          </form>
        //          <ul id='messages' class="mt-5">
        //         </ul>
                
        //      </div>
            
        //          <script>
        //              var client_id = Date.now()
        //              document.querySelector("#ws-id").textContent = client_id;
        //              var ws = new WebSocket(`ws://localhost:8000/ws/${client_id}`);
        //              ws.onmessage = function(event) {
        //                  var messages = document.getElementById('messages')
        //                  var message = document.createElement('li')
        //                  var content = document.createTextNode(event.data)
        //                  message.appendChild(content)
        //                  messages.appendChild(message)
        //              };
        //              function sendMessage(event) {
        //                  var input = document.getElementById("messageText")
        //                  ws.send(input.value)
        //                  input.value = ''
        //                  event.preventDefault()
        //              }
        //          </script>
        //      </body>
      

  


//   useEffect(() => {
//     console.log("useEffect is running")
//     // Handle connection events
//     socket.on('connect', () => {
//       setIsConnected(true);
//       console.log('Socket connected');

//       // Emit events to test functionality
//       socket.emit('join');
//       socket.emit('socket');
//     });

//     // Cleanup function to disconnect socket when component unmounts
//     return () => {
//       socket.disconnect();
//     };
//   }, [socket]);


// **************************************************************
// const Chat = () => {
//        const [userId, setUserId] = useState(2);

//         console.log("before web socket in chat.js")
//         const socket = new WebSocket(`ws://localhost:8000/ws/${userId}`);
//         console.log("socket: ", socket)
    
//         useEffect(() => {
//             socket.onopen = function (event) {
//                 console.log("WebSocket connected");
//             };
    
//             socket.onmessage = function (event) {
//                 console.log("Received message:", event.data);
//             };
    
//             socket.onclose = function (event) {
//                 console.log("WebSocket closed");
//             };
    
//             return () => {
//                 socket.close();
//             };
//         }, [socket]);
    
//         function sendEvent() {
//             const eventData = "Event data";
//             socket.send(eventData);
//             console.log("Sent event:", eventData);
//         }
    
//         return (
//             <div>
//                 <h1>WebSocket Example</h1>
//                 <button onClick={sendEvent}>Send Event</button>
//             </div>
//         );
//     };
    
//   export default Chat;




