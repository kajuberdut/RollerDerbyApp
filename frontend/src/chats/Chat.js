import React, { useState, useEffect, useRef } from "react";
import FastApi from "../Api";
import { Card, CardBody, Form, Input, Button, CardHeader, CardFooter } from "reactstrap";

function Chat({handleMessages, handleChat, chatId }) {

    const [messages, setMessages] = useState([]); 
    const [userId, setUserId] = useState();  
    const [username, setUsername] = useState(); 

    let INITIAL_STATE = { message: ""}; 
    const [formData, setFormData] = useState(INITIAL_STATE); 
    const [otherUser, setOtherUser] = useState({}); 
    const [otherUserId, setOtherUserId] = useState(); 
 
    const messagesEndRef = useRef(null); 

    const [chatParticipants, setChatParticipants] = useState([]);
    const [chatParticipantIds, setChatParticipantIds] = useState([]);
    const [groupName, setGroupName] = useState(); 

    const [isMounted, setIsMounted] = useState(false);

    /** On mount, retrieve user from local storage and set user id in state*/

    useEffect(() => {
        const user = JSON.parse(localStorage.getItem('user'));
        setUserId(user.userId);
        setUsername(user.username)
  
        if(!chatId) {
            const pathname = window.location.pathname;
            let formOtherUserId = Number(pathname.split('/')[2]);
            console.log("formOtherUserId:", formOtherUserId)
            setOtherUserId(formOtherUserId);
        }
   
      }, []);


    /** Scroll view to bottom of chat */

    const scrollToBottom = () => {
        messagesEndRef.current?.scrollIntoView({ behavior: "smooth" });
      }
  
      /** On message change call scroll to bottom */
  
      useEffect(() => {
        scrollToBottom();
      }, [messages]);

      useEffect(() => {
        // ! only running when otherUserId changes and not on MOUNT
        if (isMounted) {

            if(!chatId) {
                async function getOtherUserById() {
                    try {     
                        let otherUser =  await FastApi.getOtherUser(otherUserId);
                        setOtherUser(otherUser);
                    } catch (errors) {
                        console.error("Get Other User failed", errors);
                        return { success: false, errors };
                    }
                }
                getOtherUserById(); 
            }
            
            async function getChatHistory() {
                    try {

                    const chatHistory = chatId ? await FastApi.getChatHistoryByChatId(chatId) : await FastApi.getChatHistory([otherUserId, userId]);
                    
                    if(chatHistory) {
                      setMessages(chatHistory)
                    }
            
                    } catch (errors) {
                    console.error("Get chatHistory failed", errors);
                    return { success: false, errors };
                    }
                }
 
            getChatHistory();     

        }
        setIsMounted(true);
      }, [isMounted, otherUserId]);

      /** Format the group name to display accordingly  */

      function formatGrpNm(groupName) {
     
        let nwGrpName; 

        if(groupName.includes('& ' + username)) {
          nwGrpName = groupName.replace('& ' + username, " ")
          } 
          
        if(groupName.includes(username + ' &')) {
          nwGrpName = groupName.replace(username + ' &', " ")
        }

          if(nwGrpName) {
            return nwGrpName
          }
          return groupName
        } 

      

      useEffect(() => {
        if (isMounted) {

            if(chatId) {

                async function getChat() {
    
                /** Get chat partipants by username */
        
                try {
        
                    let chatParticipants =  await FastApi.getChatParticipants(chatId);
                    const filteredParticipants = chatParticipants.filter(
                    (participant) => participant !== username
                    );
                    const formatParticipants = filteredParticipants.join(", ");
                    setChatParticipants(formatParticipants);
        
                } catch (errors) {
        
                    console.error("Get chat particiapant failed", errors);
                    return { success: false, errors };
                }
        
                /** Get chat group name by chat id */
        
                try {
        
                    let groupName = await FastApi.getGroupNameByChatId(chatId);
                    console.log("groupName:", groupName)
                    let frmGrpNm = formatGrpNm(groupName)
                    setGroupName(frmGrpNm);
                } catch (errors) {
        
                    console.error("Get group name failed", errors);
                    return { success: false, errors };
                }
        
                /** Get chat partipant ids by chat id for socket messages  */
                
        
                    try {
                    
                    let chatParticipantIds =  await FastApi.getChatParticipantIds(chatId);
                    
                    setChatParticipantIds(chatParticipantIds)
                    } catch (errors) {
        
                    console.error("Get chatParticipantIds failed", errors);
                    return { success: false, errors };
                    }
        
                }
        
                getChat()
            }

        /** Connect to websocket  */
          FastApi.connectSocket(userId); 
  
          /** Listen for responses from websocket with event handler  */
          FastApi.socket.addEventListener("message", handleIncomingMessages);
  
          /** Remove event event listener on unmount  */
          return () => {
            FastApi.socket.removeEventListener("message", handleIncomingMessages);
            FastApi.socket.close();
          }
        }
      }, [isMounted, userId])


      function handleIncomingMessages(event) {
        console.log("Socket message received in Api.js :", event.data);
        try {
          const eventData = JSON.parse(event.data);
          console.log("!!!! event.data in chat Details!!!", event.data);
          setMessages((prevMessages) => [...prevMessages, eventData]);
        } catch (error) {
          console.error("Error parsing socket message:", error);
        }
      }


      const handleSubmit = async evt => {
        evt.preventDefault();   

        let dateTime = (new Date().toLocaleString()); 

        let messageData;

        let participantIds = chatParticipantIds ? chatParticipantIds : [otherUserId, userId];

        if(chatId !== 0) {
            messageData = {
              "messageId": 0,
              "chatId": chatId,
              "senderId": userId, 
              "senderUsername": username,
              "participantIds": participantIds,
              "message": formData.message,
              "dateTime": dateTime
            }
        } 
     
        if(!chatId) {
            console.log("hitting chatId == 0")
            messageData = {
              "messageId": 0,
              "chatId": 0,
              "participantIds": [otherUserId, userId],
              "senderId": userId, 
              "senderUsername": username,
              "message": formData.message,
              "dateTime": dateTime,
              "groupId": 0
            }
        }

        console.log("########### MESSAGE DATA ##################", messageData)
          FastApi.sendMessage(messageData)
          
          setFormData(INITIAL_STATE)     
        }

    const handleChange = evt => {
  
        const { name, value }= evt.target;
    
        setFormData(fData => ({
          ...fData,
          [name]: value,
          
        }));
      };


      return (
        <Card className="Messages"  style={{height: '500px', width: '350px', position: 'fixed', bottom: '0px', right: '480px', borderRadius: '20px'}}>
            <CardHeader style={{height: '40px'}}>

            { groupName && <p tag="h5" style={{fontWeight: 'bold', fontSize: '18px'}}>
              {groupName}
            </p>
              }


              { otherUser.username && <p style={{position: 'absolute', left: '10px', fontWeight: 'bold', fontSize: '18px'}}>{otherUser.username}</p> }

              <Button onClick={handleChat ? handleChat : handleMessages} style={{ position: 'absolute', right: '4px', top: '0',  backgroundColor: 'transparent', color: 'black', border: 'none', fontSize: '18px'  }}>X</Button>
            </CardHeader>

            <CardBody>

                {messages &&
                <div style={{ overflowY: 'auto', overflowX: 'hidden', height: 300 }}> 
                          {messages.map((message) => (
                        
                        <div key={'history' + message.messageId}>
                            
                          <p style={{textAlign: 'left'}}>
                            <span style={{ display: 'inline-block', paddingLeft: message.userId == userId ? '80px' : '0px', fontSize: 'smaller', fontStyle: 'italic' }}>
                              {message.senderUsername}
                            </span>
                          </p>
                          
                        
                          <p style={{ backgroundColor: message.userId == userId ? 'lightgray' : 'lightblue', borderRadius: '10px', margin: '5px', padding: '5px', width: '210px', marginLeft: message.userId == userId ? '80px' : '0px', textAlign: 'left' }}>

                          {message.message} 
                            <br />
                      
  
                            <span style={{ visibility: 'hidden' }} ref={messagesEndRef}></span>
  
                          </p>
                        </div>
                      
                  ))}
                </div> 
                }    
            </CardBody>
            <CardFooter>
                {/* <Form style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', height: '50px'}} onSubmit={handleSubmit}> */}
                <Form style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', height: '50px'}}>
                    <Input
                      type="textarea"
                      id="message"
                      name="message"
                      value={formData.message}
                      onChange={handleChange}
                      placeholder="Type a message.."
                      style={{ width: '240px', height: '40px', resize: 'none', bottom: '7px'}}
                      onKeyDown={(evt) => {
                        if (evt.key === 'Enter') {
                          handleSubmit(evt);
                        }
                      }}
                    />
                    <Button onClick={handleSubmit} style={{width: '63px', height: '36px'}}>
                        Send
                    </Button>
                 </Form>
            </CardFooter>
        </Card>        
      )

    }; 

    export default Chat;
