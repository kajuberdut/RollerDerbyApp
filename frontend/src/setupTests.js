// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';


global.localStorage = {
  getItem: jest.fn(),
  setItem: jest.fn(),
};

const user = {
  userId: 1,
  username: "testUser",
  email: "testuser@gmail.com",
  facebookName: "facebook test user",
  about: "This is my story...",
  primaryNumber: "01",
  level: "A",
  ruleset: [],
  position: [],
  insurance: [],
  locationId: null,
  associatedLeagues: "Testing League",
  phoneNumber: "1111111111",
  firstName: "Jane",
  lastName: "Doe",
  additionalInfo: "Epi pen in bag.",
  secondaryNumber: "02"
};

beforeEach(() => {
  localStorage.setItem("user", JSON.stringify(user));
});

