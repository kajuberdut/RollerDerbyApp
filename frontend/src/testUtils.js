import React from "react";
import UserContext from "./multiUse/UserContext";

const demoUser = {
    userId: 1,
    username: "testUser",
    email: "testuser@gmail.com",
    facebookName: null,
    about: null,
    primaryNumber: null,
    level: null,
    ruleset: [],
    position: [],
    insurance: [],
    locationId: null,
    associatedLeagues: null,
    phoneNumber: null,
    firstName: null,
    lastName: null,
    additionalInfo: null,
    secondaryNumber: null
  }

const UserProvider =
    ({ children, user = demoUser }) => (
    <UserContext.Provider value={{ user }}>
      {children}
    </UserContext.Provider>
);



export { UserProvider };
