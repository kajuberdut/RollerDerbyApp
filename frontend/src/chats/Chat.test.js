import { render, screen } from '@testing-library/react';
import Chat from './Chat';
import { UserProvider } from "../testUtils";
import { BrowserRouter } from "react-router-dom";
import '../setupTests.js';

// * Passing as of 2/13/24
// npm test Chat.test.js
// must be in frontend directory  

describe("Chat Page Render", () => {

  test('renders chat page', () => {
    
    render(
    <BrowserRouter>
            <Chat />
    </BrowserRouter>);

  });


  test('matches snapshot', function() {
    const { asFragment } = render(<Chat />);
    expect(asFragment()).toMatchSnapshot(); 
  })

})

describe("Chat Page Contents", () => {

  test('renders chat page', () => {
    
    render(
    <BrowserRouter>
            <Chat />
    </BrowserRouter>);

    const button = screen.getByText('Send'); 
    expect(button).toBeInTheDocument();

  });

})