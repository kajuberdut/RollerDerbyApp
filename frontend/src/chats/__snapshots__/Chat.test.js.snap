// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Page Render matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Messages card"
    style="height: 500px; width: 350px; position: fixed; bottom: 0px; right: 480px; border-radius: 20px;"
  >
    <div
      class="card-header"
      style="height: 40px;"
    >
      <button
        class="btn btn-secondary"
        style="position: absolute; right: 4px; top: 0px; background-color: transparent; color: black; font-size: 18px;"
      >
        X
      </button>
    </div>
    <div
      class="card-body"
    >
      <div
        style="overflow-y: auto; overflow-x: hidden; height: 300px;"
      />
    </div>
    <div
      class="card-footer"
    >
      <form
        class=""
        style="display: flex; justify-content: space-between; align-items: center; height: 50px;"
      >
        <textarea
          class="form-control"
          id="message"
          name="message"
          placeholder="Type a message.."
          style="width: 240px; height: 40px; resize: none; bottom: 7px;"
        />
        <button
          class="btn btn-secondary"
          style="width: 63px; height: 36px;"
          type="button"
        >
          Send
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Messages card"
    style="height: 500px; width: 350px; position: fixed; bottom: 0px; right: 480px; border-radius: 20px;"
  >
    <div
      class="card-header"
      style="height: 40px;"
    >
      <button
        class="btn btn-secondary"
        style="position: absolute; right: 4px; top: 0px; background-color: transparent; color: black; font-size: 18px;"
      >
        X
      </button>
    </div>
    <div
      class="card-body"
    >
      <div
        style="overflow-y: auto; overflow-x: hidden; height: 300px;"
      />
    </div>
    <div
      class="card-footer"
    >
      <form
        class=""
        style="display: flex; justify-content: space-between; align-items: center; height: 50px;"
      >
        <textarea
          class="form-control"
          id="message"
          name="message"
          placeholder="Type a message.."
          style="width: 240px; height: 40px; resize: none; bottom: 7px;"
        />
        <button
          class="btn btn-secondary"
          style="width: 63px; height: 36px;"
          type="button"
        >
          Send
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
